version-negotiation.md

```diff
@@ -0,0 +1,74 @@
+
+### Privacy-preserving
+
+Any scenario in which the browser communicates previously-sent information back to the server or website can be (ab)used as a cookie. Previous incarnations of origin policy communicated the currently-downloaded policy version. But even the behaviors controlled by origin policy could be used in this manner.
+
+## Proposed model
+
+Conceptually, origin policies are stored in the HTTP cache, under the URL `$origin/.well-known/origin`. In particular, this means they are double-keyed [in the same way as the HTTP cache](https://github.com/WICG/origin-policy/issues/23): that is, by (origin of document, origin of top-document). This prevents them from being used for cross-site tracking.
+
+(Question: should clearing the HTTP cache also clear any cached origin policy? Probably yes. Or should it be tied to clearing cookies?)
+
+When the browser makes a request to _url_, it:
+
+* Checks if it has something cached, and not expired, for `$that_origin/.well-known/origin-policy`. If so, this is the _candidate origin policy_; otherwise the _candidate origin policy_ is the "none" policy.
+* Makes the request to _url_.
+* The response can contain a header of the form `Origin-Policy: allowed=(policyA polB polC none), preferred=policyZ`. This indicates that policies with identifiers `policyA`, `polB`, `polC`, or `policyZ` are acceptable to the site, or no origin policy at all. Call these the _list of acceptable policies_. If the response contains no such header, then the _list of acceptable policies_ contains just the "none" policy.
+* Checks the _list of acceptable policies_ against the _candidate origin policy_.
+  * If the _candidate origin policy_ is in the _list of acceptable origin policies_, then:
+    * Apply the _candidate origin policy_ and load the response. (This might apply the "none" policy.)
+    * If _candidate origin policy_ is not the preferred origin policy (indicated by the `preferrered=` portion), then the browser sends a low-priority request to `$that_origin/.well-known/origin-policy` to refresh the HTTP cache, but it won't apply for this page load.
+  * Otherwise, if the _list of acceptable origin policies_ only contains the "none" policy but _candidate origin policy_ is not the none policy, then apply the "none" policy anyway, and load the response.
+  * Otherwise, the browser makes a request (on the same connection, if HTTP/2), to `$that_origin/.well-known/origin-policy`. It delays any loading of the response for _url_ until the new policy has been loaded. If the new policy's identifier still doesn't match the _list of acceptable policies_, then the result of the origin navigation request is a network error.
+
+Here, the identifier for an origin policy is found inside its JSON document, e.g. `"identifier": "policyA"`. If the value is `"none"` that is treated as a parse error. (Maybe we should create a space of reserved identifiers and change `"none"` to one of them? E.g. `"__none__"`?)
```

Comment:
"none" as parse error: This is unlikely to be a good idea. The current choice is that a parse error is treated as a fatal error, which essentially "bricks" the site. (Actually, a scary click-through interstitial, but still.) Bricking a site due to config errors is unpopular. I'd strongly prefer some form of error handling that distinguished between 'hard' and ignorable errors, and to keep the set of 'hard' small.

Reply:
Could you expand on this potential distinction more? How is using a reserved policy name (like `"none"` or `"__none__"`) different from forgetting a comma in your JSON? Or are you thinking that JSON parse errors should also fail-soft?

Reply:
In our current implementation, we implement rules vaguely like:
- JSON syntax error => hard error
- unknown top-level key => ignore.
- parse-error in a known top-level key => hard error.
- retrieval error for an announced policy (and no cache hit) => hard error

This isn't any strict criteria, unfortunately. I do suspect error handling (for the entire feature) will be difficult: Security-minded people seem to prefer hard errors; but that makes deployment risky. allowed=(none ...) does make that easier.

Reply:
Currently I feel that using a reserved policy name is in the same category as a missing comma, and would like to treat them the same. We should indeed have a more general discussion about error handling as the model settles, but for now I'll leave this in.